crates/core/plugin_sm/src/plugin.rs

```diff
@@ -461,13 +468,54 @@ impl Plugin for ExternalPluginCommand {
         let command = self.command(LIST, None)?;
         let output = self.execute(command, logger).await?;
         if output.status.success() {
+            let filtered_output = match (&self.exclude, &self.include) {
+                (None, None) => output.stdout,
+                (Some(exclude), Some(include)) => {
+                    let exclude_filter = Regex::new(format!(r"^{}\t.*", exclude).as_str())?;
+                    let include_filter = Regex::new(format!(r"^{}\t.*", include).as_str())?;
+
+                    output
+                        .stdout
+                        .split_inclusive(|c| *c == b'\n')
+                        .map(|line| std::str::from_utf8(line).unwrap_or_default())
+                        .filter(|line| {
+                            include_filter.is_match(line) || !exclude_filter.is_match(line)
+                        })
+                        .flat_map(|line| line.as_bytes().to_vec())
+                        .collect()
+                }
+                _ => {
+                    let exclude_filter = if let Some(exclude) = self.exclude.as_ref() {
+                        Regex::new(format!(r"^{}\t.*", exclude).as_str())?
+                    } else {
+                        Regex::new(r"^$")?
+                    };
+
+                    let include_filter = if let Some(include) = self.include.as_ref() {
+                        Regex::new(format!(r"^{}\t.*", include).as_str())?
+                    } else {
+                        Regex::new(r".*")?
+                    };
+
+                    output
+                        .stdout
+                        .split_inclusive(|c| *c == b'\n')
+                        .map(|line| std::str::from_utf8(line).unwrap_or_default())
+                        .filter(|line| {
+                            include_filter.is_match(line) && !exclude_filter.is_match(line)
+                        })
+                        .flat_map(|line| line.as_bytes().to_vec())
+                        .collect()
+                }
+            };
```

Comment:
suggestion: Is there a specific reason why we do "both filters present" and "single filter present" cases slightly differently? If there's not, I think writing it in terms of a single, general case, would be easier to follow, as right now it's a bit confusing, understanding why we use fallbacks that we do and why the filter condition is OR first and AND later.

Reply:
No, we fell into the trap before. When both inclusive and exclusive filters are applied, then it is an OR condition. For example if a users has an exclude filter, but they want to have an exception to the exclude, then they can use a single "include" pattern to specify the exceptions.

Reply:
Yes, when two filters are present, the behavior is different (thus implementation is slightly different) as described in Didier's comment above. Basically, include filter takes precedence over exclude one in that case.

Reply:
By the way, the *and clause* in  the `_` case is not really an *and* as there is only a single regex that is user provided.

@Ruadhri17 I think you can fallback to your very fist proposal with 4 explicit cases: this mix of `||` and `&&` with puzzle any new comers.

Reply:
from [Didier's comment](https://github.com/thin-edge/thin-edge.io/pull/2869#issuecomment-2112298883) (emphasis mine):

- if both are provided, the logic is to keep a package if its name matches the software.plugin.include regex or doesn't match the software.plugin.exclude. The logic is that the primary usage will be to exclude packages (say all libs and kernel modules) and only then add some exceptions (say to include the libc).
- if only software.plugin.include is provided, keep a package if its name matches **(and deny everything else)**
- if only software.plugin.exclude is provided, keep a package if its name doesn't match

I have trouble understanding why "if only `include` is provided then we only keep things it matches" doesn't directly map into "both `include` and `exclude` are provided but `exclude` denies everything" case.

> When both inclusive and exclusive filters are applied, then it is an OR condition. For example if a users has an exclude filter, but they want to have an exception to the exclude, then they can use a single "include" pattern to specify the exceptions.

Yeah, I'm with you on that one. But I don't understand why do we need the second AND case. If we only have `include`, and only want to keep items matched by `include`, I think we can use a fallback value of "deny everything" for `exclude` and reuse the OR condition. Alternatively, if we only have `exclude` given by the user, then for include we can use a regex that includes nothing (so it doesn't add any exceptions) and again we would be able to use the OR condition.

So if I'm wrong, and I'm stepping into the same trap than you did (because maybe there's some hidden assumption that doesn't follow from these 3 rules), then I think we should be extra careful how this would be documented.

Reply:
After quick discussion with @Bravo555 we came to agreement that when you switch default filters, only two cases are needed:

1.  When both filters are None - we just pass the output list and skip unnecessary. Switched filters would not produce desired output anyway.
2. All other cases can be covered by OR condition.